x-pack/plugins/enterprise_search/server/lib/indices/text_analysis.ts

```diff
@@ -0,0 +1,298 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+interface LanguageDataEntry {
+  name: string;
+  stemmer: string;
+  stop_words: string;
+  custom_filter_definitions?: object;
+  prepended_filters?: string[];
+  postpended_filters?: string[];
+}
+
+const LanguageData: Record<string, LanguageDataEntry> = {
+  da: {
+    name: 'Danish',
+    stemmer: 'danish',
+    stop_words: '_danish_',
+  },
+  de: {
+    name: 'German',
+    stemmer: 'light_german',
+    stop_words: '_german_',
+  },
+  en: {
+    name: 'English',
+    stemmer: 'light_english',
+    stop_words: '_english_',
+  },
+  es: {
+    name: 'Spanish',
+    stemmer: 'light_spanish',
+    stop_words: '_spanish_',
+  },
+  fr: {
+    name: 'French',
+    stemmer: 'light_french',
+    stop_words: '_french_',
+    custom_filter_definitions: {
+      'fr-elision': {
+        type: 'elision',
+        articles: [
+          'l',
+          'm',
+          't',
+          'qu',
+          'n',
+          's',
+          'j',
+          'd',
+          'c',
+          'jusqu',
+          'quoiqu',
+          'lorsqu',
+          'puisqu',
+        ],
+        articles_case: true,
+      },
+    },
+    prepended_filters: ['fr-elision'],
+  },
+  it: {
+    name: 'Italian',
+    stemmer: 'light_italian',
+    stop_words: '_italian_',
+    custom_filter_definitions: {
+      'it-elision': {
+        type: 'elision',
+        articles: [
+          'c',
+          'l',
+          'all',
+          'dall',
+          'dell',
+          'nell',
+          'sull',
+          'coll',
+          'pell',
+          'gl',
+          'agl',
+          'dagl',
+          'degl',
+          'negl',
+          'sugl',
+          'un',
+          'm',
+          't',
+          's',
+          'v',
+          'd',
+        ],
+        articles_case: true,
+      },
+    },
+    prepended_filters: ['it-elision'],
+  },
+  ja: {
+    name: 'Japanese',
+    stemmer: 'light_english',
+    stop_words: '_english_',
+    postpended_filters: ['cjk_bigram'],
+  },
+  ko: {
+    name: 'Korean',
+    stemmer: 'light_english',
+    stop_words: '_english_',
+    postpended_filters: ['cjk_bigram'],
+  },
+  nl: {
+    name: 'Dutch',
+    stemmer: 'dutch',
+    stop_words: '_dutch_',
+  },
+  pt: {
+    name: 'Portuguese',
+    stemmer: 'light_portuguese',
+    stop_words: '_portuguese_',
+  },
+  'pt-br': {
+    name: 'Portuguese (Brazil)',
+    stemmer: 'brazilian',
+    stop_words: '_brazilian_',
+  },
+  ru: {
+    name: 'Russian',
+    stemmer: 'russian',
+    stop_words: '_russian_',
+  },
+  th: {
+    name: 'Thai',
+    stemmer: 'light_english',
+    stop_words: '_thai_',
+  },
+  zh: {
+    name: 'Chinese',
+    stemmer: 'light_english',
+    stop_words: '_english_',
+    postpended_filters: ['cjk_bigram'],
+  },
+};
+
+const FRONT_NGRAM_MAX_GRAM = 12;
+
+const edgeEndgramType = 'edge_ngram' as 'edge_ngram';
```

Comment:
I think you can try to set a type for 'GenericFilters' explicitly. I am guessing your issues coming from TS trying to infer types, and infer these as "string" instead of literals that you want.

Reply:
I checked locally, and can confirm, that currently TypeScript infers `GenericFilters` type from the shape. If you use `elasticsearch-js` types you can get rid of this issue.
![Screenshot 2022-07-08 at 14 15 20](https://user-images.githubusercontent.com/1410658/177990420-c53a2436-20f1-4acc-ac1b-fd481a9d5317.png)

So to test it out, I put `Record<string, AnalysisTokenFilter>` as a type for generic filters. You can see below how the `type` changed in between two. (I deleted the type casts of course)
![Screenshot 2022-07-08 at 14 17 28](https://user-images.githubusercontent.com/1410658/177990832-41ec0628-4625-4053-b813-10a7928b24cb.png)
![Screenshot 2022-07-08 at 14 17 41](https://user-images.githubusercontent.com/1410658/177990850-8df0f961-7782-4d8a-a62d-977df882425d.png)

Reply:
Efe is exactly right. As a side-note: not a big fan of how they designed that type tbh.

Reply:
Thanks! With this and a little of @yakhinvadim's help, I was able to refactor into a somewhat cleaner solution. I've pushed up the changes.